File: stream/java/src/com/db/stream/DataMutator.java
/*
 * Copyright (c) 2007 Digital Bazaar, Inc.  All rights reserved.
 */
package com.db.stream;

import java.io.IOException;
import java.io.InputStream;

import com.db.util.ByteBuffer;

/**
 * A DataMutator mutates data that is stored in a ByteBuffer and writes it to
 * another ByteBuffer from which it can be retrieved.
 * 
 * @author Dave Longley
 */
public class DataMutator
{
   /**
    * The source ByteBuffer with unmutated data.
    */
   protected ByteBuffer mSource;
   
   /**
    * The destination ByteBuffer with mutated data.
    */
   protected ByteBuffer mDestination;
   
   /**
    * The algorithm used to mutate data.
    */
   protected DataMutationAlgorithm mAlgorithm;
   
   /**
    * Creates a new DataMutator.
    * 
    * @param src the source ByteBuffer to read from.
    * @param dest the destination ByteBuffer to write to.
    */
   public DataMutator(ByteBuffer src, ByteBuffer dest)
   {
      // store buffers
      mSource = src;
      mDestination = dest;
      
      // no algorithm yet
      mAlgorithm = null;
   }
   
   /**
    * Sets the data mutation algorithm for this mutator.
    * 
    * @param algorithm the DataMutationAlgorithm to use.
    */
   public void setAlgorithm(DataMutationAlgorithm algorithm)
   {
      mAlgorithm = algorithm;
   }
   
   /**
    * Reads data from the passed input stream and mutates it.
    * 
    * This method will block until at one byte can be read from the input
    * stream, unless this source buffer for this mutator is already full, or
    * until the end of the stream has been reached.
    * 
    * @param is the input stream to read from.
    * 
    * @return the number of bytes read from the input stream and put into
    *         this mutator, or -1 if the end of the input stream was reached.
    * 
    * @exception IOException thrown if an IO error occurs.
    */
   public int put(InputStream is) throws IOException
   {
      int rval = mSource.put(is);
      
      // only mutate data if the destination is empty
      if(mDestination.isEmpty())
      {
         if(mAlgorithm != null)
         {
            // mutate the data
            mAlgorithm.mutateData(mSource, mDestination);
         }
         else
         {
            // get all data from source and put it in destination
            mSource.get(mDestination, mSource.getUsedSpace(), true);
         }
      }
      
      return rval;
   }
   
   /**
    * Gets data out of this mutator and puts it into the passed buffer. The
    * amount of data may be less than the requested amount if this mutator
    * does not have enough data.
    * 
    * @param b the buffer to put the retrieved data into.
    * @param offset the offset at which to put the retrieved data.
    * @param length the maximum number of bytes to get.
    * 
    * @return the actual number of bytes retrieved, which may be 0 if this
    *         mutator is empty.
    */
   public int get(byte[] b, int offset, int length)
   {
      return mDestination.get(b, offset, length);
   }
   
   /**
    * Returns true if this mutator has data that can be obtained, false if
    * not.
    * 
    * @return true if this mutator has data that can be obtained, false if
    *         not.
    */
   public boolean hasData()
   {
      return !mDestination.isEmpty();
   }
}

File: io/cpp/db/io/DataMutator.cpp
/*
 * Copyright (c) 2007-2008 Digital Bazaar, Inc.  All rights reserved.
 */
#include "db/io/DataMutator.h"

using namespace db::io;

DataMutator::DataMutator(ByteBuffer* src, ByteBuffer* dest)
{
   // store buffers
   mSource = src;
   mDestination = dest;
   
   // no algorithm yet, use default
   mAlgorithm = this;
   mAlgorithmExitCode = 0;
}

DataMutator::~DataMutator()
{
}

int DataMutator::mutateData(ByteBuffer* src, ByteBuffer* dest, bool finish)   
{
   int rval = 0;
   
   if(mAlgorithm == this)
   {
      if(!src->isEmpty())
      {
         // get all data from source and put it in destination
         src->get(dest, src->length(), true);
         rval = 1;
      }
      
      if(finish)
      {
         // algorithm finished
         mAlgorithmExitCode = rval = 2;
      }
   }
   else
   {
      // use set algorithm
      rval = mAlgorithm->mutateData(src, dest, finish);
   }
   
   return rval;
}

void DataMutator::setAlgorithm(DataMutationAlgorithm* algorithm)
{
   mAlgorithm = algorithm;
   mAlgorithmExitCode = 0;
}

int DataMutator::mutate(InputStream* is)
{
   int rval = 0;
   
   // mutate while no data is available and not finished
   bool finish = false;
   bool read = false;
   while(rval != -1 && !hasData() && mAlgorithmExitCode == 0)
   {
      // read as necessary
      if(read)
      {
         rval = mSource->put(is);
         finish = (rval == 0);
      }
      
      if(rval != -1)
      {
         // try to mutate data
         rval = mAlgorithm->mutateData(mSource, mDestination, finish);
         
         if(rval > 1)
         {
            // algorithm completed
            mAlgorithmExitCode = rval;
         }
         else
         {
            // read if mutation algorithm needs more data
            read = (rval == 0);
         }
      }
   }
   
   // handle remaining source data if algorithm has completed
   if(rval != -1 && mAlgorithmExitCode > 0)
   {
      if(mAlgorithmExitCode == 2)
      {
         // copy any source data to destination buffer
         mSource->get(mDestination, mSource->length(), true);
         
         // fill destination buffer directly
         rval = mDestination->put(is);
      }
      else if(mAlgorithmExitCode == 3)
      {
         // truncate remaining data
         mSource->clear();
         while((rval = mSource->put(is)) > 0)
         {
            mSource->clear();
         }
      }
   }
   
   if(rval > 0)
   {
      rval = mDestination->length();
   }
   
   return rval;
}

int DataMutator::mutate(const char* b, int length)
{
   int rval = 0;
   
   if(mAlgorithmExitCode == 0)
   {
      // try to finish algorithm if length is zero
      bool finish = (length == 0);
      
      // determine if there is any cached source data
      if(!mSource->isEmpty())
      {
         // append the data to the source buffer
         mSource->put(b, length, true);
         
         // try to mutate data
         rval = mAlgorithm->mutateData(mSource, mDestination, finish);
      }
      else
      {
         // try to mutate passed data without caching it
         mInputWrapper.setBytes((char*)b, 0, length, false);
         rval = mAlgorithm->mutateData(&mInputWrapper, mDestination, finish);
         
         // copy excess bytes into source buffer
         if(!mInputWrapper.isEmpty())
         {
            mSource->put(&mInputWrapper, mInputWrapper.length(), true);
         }
      }
   }
   
   if(rval != -1)
   {
      if(mAlgorithmExitCode == 2)
      {
         // copy data to destination
         mSource->get(mDestination, mSource->length(), true);
         mDestination->put(b, length, true);
      }
      else if(mAlgorithmExitCode == 3)
      {
         // truncate data
         mSource->clear();
         mDestination->clear();
      }
   }
   
   if(rval > 0)
   {
      rval = mDestination->length();
   }
   
   return rval;
}

long long DataMutator::skipMutatedBytes(InputStream* is, long long count)
{
   long long rval = 0;
   
   // mutate and skip data
   long long remaining = count;
   while(remaining > 0 && mutate(is))
   {
      // clear bytes
      remaining -= mDestination->clear((int)(remaining & 0x7fffffff));
   }
   
   if(remaining < count)
   {
      // some bytes were skipped
      rval = count - remaining;
   }
   
   return rval;
}

int DataMutator::get(char* b, int length)
{
   return mDestination->get(b, length);
}

bool DataMutator::hasData()
{
   return !mDestination->isEmpty();
}

void DataMutator::setSource(ByteBuffer* src)
{
   mSource = src;
}

ByteBuffer* DataMutator::getSource()
{
   return mSource;
}

void DataMutator::setDestination(ByteBuffer* dest)
{
   mDestination = dest;
}

ByteBuffer* DataMutator::getDestination()
{
   return mDestination;
}
